### DISCUSSION TITLE
Search for "hex-string"

### DISCUSSION THREAD

**Discussion by GflowerG** [dnGrep/dnGrep#532] [Ideas]
Why not make it possible to search for raw content?

More specifically, similar to the Total commander (search files dialog) application - the "Hex" checkbox. Just add another "HEX" option to the existing "Regex" "XPath" "Text" or "Phonetic".

### Possible query syntaxes:
```
48 65 6C 6C 6F 21 00 00

\x48\x65\x6C\x6C\x6F\x21\x00\x00

48656C6C6F210000
```

### Wildcards examples:
```
48 65 6C 6C ?? ?? 00 00

48 65 6C 6C ** ** 00 00
```

etc...

**Comment by doug24** [comment]
I could see doing something like this: read the file as binary, and convert each byte as pair of hex numbers as a string, and then search the resulting text.  Implementing the wildcards wouldn't be easy, and it may be easier to make them into a regular expression and do a regex search.  Would the ? and * wildcards do something different in your example?

The biggest problem I see is showing the results.  The dnGrep controls show lines of text, and there aren't any line breaks in the hex data.  You couldn't open more than a very small file in the search results pane.  The preview pane shows files, so dnGrep would have to write a temp file of the binary-to-hex-string data to open it in the preview pane.  But I think dnGrep would have to break the byte data up into lines for displaying the results - I'm thinking of something like what VIM does when you convert to hex in that editor.  There's some work there...

**Reply by GflowerG** [reply]
Wildcards are not the most requested feature.

But a simple search for hex arrays would be helpful. (48 65 6C 6C 6F 21 00 00)

**Reply by GflowerG** [reply]
I have one more question...

Can you improve the display behavior of the results? I mean that it is better to display not a whole line, but a user-specified offset of characters from the found phrase.

For example, if there is a string "123 456 789", I will search for "456". I'll set the offsets to -3 and +2.

Now the result is like this:
"123 **456** 789".

My idea:
"23 **456** 7" (-3 ; +2)

It's better for performance and readability.

**Reply by doug24** [reply]
I'm going to bet that nearly everyone will think that is not an improvement, and will they will want context that is at least one line, and most will have few lines before/after.  So I do not want to change the existing behavior from lines to characters.  But maybe an option to specify if the context count is in lines or characters.  I'll have to see how much work it will be to add it.

**Reply by GflowerG** [reply]
I am using your tool to find unicode strings in binaries (portable executable).

Even if I turn off the "before and after" line count, some of the lines are very large and long.

This character offset idea - fixes the problem.

And if I understood you correctly, this algorithm would solve this problem too:
> The biggest problem I see is showing the results. The dnGrep controls show lines of text, and there aren't any line breaks in the hex data. You couldn't open more than a very small file in the search results pane.
